src/borg/archiver.py

```diff
@@ -1099,6 +1103,113 @@ def print_output(diff, path):
+                
+                if isinstance(value, bytes):
+                    if len(value) < 65:
+                        return bin_to_hex(value)
+                    else:
+                        return f'{len(value)} bytes'
+
+                if isinstance(value, list):
+                    return [ fixup_json(v) for v in value ]
+
+                return value
+
+            def dump_item(id, data):
+                try:
+                    msg = msgpack.unpackb(data)
+                except UnpackException:
+                    msg = { b'data': data }
+
+                msg = fixup_json(msg)
+                msg['id'] = bin_to_hex(id)
+                json_print(msg)
+        else:
+            packer = msgpack.Packer()
+            def dump_item(id, data):
+                sys.stdout.buffer.write(packer.pack({
+                    b'id': id,
+                    b'data': data,
+                }))
+
+        marker = None
+        while True:
+            result = repository.scan(limit=LIST_SCAN_LIMIT, marker=marker)
+            if not result:
+                break
+            marker = result[-1]
+            for id in result:
+                cdata = repository.get(id)
+                give_id = id if id != Manifest.MANIFEST_ID else None
+                data = key.decrypt(give_id, cdata)
+                dump_item(id, data)
+
+        return self.exit_code
+
+    @with_repository(exclusive=True, compatibility=(Manifest.Operation.WRITE,))
+    def do_load(self, args, repository, manifest, key):
+        """load repository contents from stdin"""
+
+        # set up the key without depending on a manifest obj
+        # ids = repository.list(limit=1, marker=None)
+        # cdata = repository.get(ids[0])
+        # key = key_factory(repository, cdata)
+        
+        def add_archive(id, data):
+            try:
+                raw = msgpack.unpackb(data)
+            except msgpack.UnpackException:
+                return
+
+            if isinstance(raw, dict) and b'name' in data and b'version' in data:
+                archive = ArchiveItem(internal_dict=raw)
+                logger.info('Found archive %s', archive.name)
+                manifest.archives[archive.name] = (id, archive.time)
+
+        unpacker = msgpack.Unpacker(file_like=sys.stdin.buffer, object_hook=StableDict)
+        while True:
+            try:
+                item = unpacker.unpack()
+            except msgpack.OutOfData:
+                print('OUT OF DATA')
+                break
+            
+            if item[b'id'] == Manifest.MANIFEST_ID:
+                print('SKIP MANIFEST')
+                continue
+
+            id_ = bin_to_hex(key.id_hash(item[b'data']))
+            print(id_[:10], bin_to_hex(item[b'id'])[:10])
+            cdata = key.encrypt(item[b'data'])
+            repository.put(item[b'id'], cdata)
+
+            add_archive(item[b'id'], item[b'data'])
```

Comment:
a less magic approach would be:
- at the origin repo (when dumping), transmit a set of all archive IDs (or just the full manifest?) **first**
- at the destination repo (when loading) check if the current ID is in that set and then do sth like in add_archive.

Reply:
I agree that it would be better for the origin to identify the archives, though instead of a list I was thinking to tag the archive packets.